File: modules/wem-webapp/src/main/webapp/admin/api/js/schema/content/form/InputTypeName.ts
module api_schema_content_form{

    export class InputTypeName {

        private static CUSTOM_PREFIX:string = "custom:";

        private custom:boolean;

        private name:string;

        private refString:string;

        static parseInputTypeName(str:string) {
            if (str.substr(0, InputTypeName.CUSTOM_PREFIX.length) == InputTypeName.CUSTOM_PREFIX) {
                return new InputTypeName(str.substr(InputTypeName.CUSTOM_PREFIX.length, str.length), true);
            }
            else {
                return new InputTypeName(str, false);
            }
        }

        constructor(name:string, custom:boolean) {
            this.name = name;
            this.custom = custom;

            if( this.custom ) {
                this.refString = InputTypeName.CUSTOM_PREFIX + name;
            }
            else {
                this.refString = name;
            }
        }

        getName():string {
            return this.name;
        }

        isBuiltIn():boolean {
            return !this.custom;
        }

        toString():string {
            return this.refString;
        }
    }
}

File: modules/core-api/src/main/java/com/enonic/wem/api/form/inputtype/InputTypeName.java
package com.enonic.wem.api.form.inputtype;


import java.util.Objects;

public final class InputTypeName
{
    private final String ref;

    private final String name;

    private final boolean custom;

    public static InputTypeName from( final String s )
    {
        if ( s.startsWith( "custom:" ) )
        {
            return new InputTypeName( s.substring( "custom:".length(), s.length() ), true );
        }
        else
        {
            return new InputTypeName( s, false );
        }
    }

    public static InputTypeName from( final InputType inputType )
    {
        return InputTypeName.from( inputType.getName() );
    }

    public InputTypeName( final String name, final boolean custom )
    {
        this.name = name;
        this.custom = custom;
        this.ref = custom ? "custom:" + name : "" + name;
    }

    public String getName()
    {
        return this.name;
    }

    public boolean isCustom()
    {
        return this.custom;
    }

    @Override
    public boolean equals( final Object o )
    {
        if ( this == o )
        {
            return true;
        }
        if ( !( o instanceof InputTypeName ) )
        {
            return false;
        }

        final InputTypeName other = (InputTypeName) o;

        return Objects.equals( ref, other.ref ) && Objects.equals( ref, other.ref );
    }

    @Override
    public int hashCode()
    {
        return Objects.hash( ref );
    }

    public String toString()
    {
        return ref;
    }
}
